### ISSUE TITLE
[RFC] Checkpoint Integrity Verification (SHA-256 Manifest)

### ISSUE THREAD

**Issue by chango9543** [pytorch/torchtitan#4011]
# RFC: Checkpoint Integrity Verification (SHA-256 Manifest)

## Summary

Add an opt-in integrity verification mechanism for DCP and HuggingFace
safetensors checkpoints. On save, write a per-step manifest of file hashes;
on load, recompute hashes and reject the load if any file has been silently
corrupted (disk bit-rot, truncated writes, NFS transfer errors, ...).

## Motivation

Large-scale distributed training (hundreds of cards, multi-day runs) treats
checkpoints as the only durable recovery point. Today torchtitan's
`CheckpointManager` considers a step "valid" as long as `.metadata` or
`model.safetensors.index.json` exists (see `_find_load_step`). There is
**no content-level integrity check**, so:

- A single bit flip on disk (silent data corruption, well documented on
  large NFS/Ceph stores) loads corrupt weights into the model. Loss does
  not blow up immediately - it silently drifts, and the corruption is
  discovered days later when metrics diverge for unexplained reasons.
- Truncated writes from a crash mid-save leave a half-written `.distcp`
  file. `_find_load_step` still picks it up because `.metadata` exists.
- HF safetensors exported for downstream inference/republishing ship
  corrupt weights to consumers.

PyTorch DCP and safetensors both lack an end-to-end content hash. This
gap is also acknowledged by Megatron-LM, which recently added
`verify_integrity` + `integrity.json` (see `megatron/core/dist_checkpointing/validation.py`).

## Goals

- Detect accidental checkpoint corruption (bit rot, truncation, transfer
  errors) before any weight reaches the model.
- Work for **both** DCP (`.metadata` + `__*.distcp`) and HuggingFace
  safetensors layouts without layout-specific code.
- Stay **opt-in** so existing checkpoints and CI loops are unaffected.
- Be **algorithm-pluggable** so users can trade security vs. speed
  (sha256 / crc32) without touching core code.
- Distributed-safe: only rank 0 reads the files; the verdict is
  broadcast so all ranks raise in lock-step or proceed together.

### Manifest layout

Each `step-N/` directory gains a small JSON manifest, e.g.
`_integrity_manifest.json`:

```json
{
  "algorithm": "sha256",
  "files": {
    ".metadata": "e3b0c44298fc1c14...",
    "__0_0.distcp": "9e86d671..."
  }
}
```

- The manifest is excluded from its own hash (chicken-and-egg).
- Files are sorted by name so the manifest is byte-identical across runs
  and ranks - important for diff-based CI.
- `algorithm` is recorded in the manifest so loading picks the algorithm
  from the manifest, not the current config. A sha256-written manifest
  can be verified years later even if the default algorithm changed.

### Save path

After `dcp.save()` / `consolidate_safetensors_files_on_every_rank()`
returns (i.e., the on-disk layout is final), rank 0 walks the step
directory, hashes every file, and writes the manifest. A `barrier()`
follows so no rank starts loading before the manifest is visible.

For `AsyncMode.ASYNC` / `ASYNC_WITH_PINNED_MEM`, the manifest write is
registered as a `add_done_callback` on the upload future, with a
blocking `.result()` fallback for future types that lack callbacks
(Megatron uses the same `finalize_fn` pattern - see
`serialization.py`).

### Load path

Before `dcp.load()` / `safe_open()`, call `verify_manifest_distributed`:

1. If `verify_integrity` is False -> skip (zero overhead).
2. Rank 0 recomputes each file's hash using the algorithm recorded in
   the manifest, compares, and collects **all** mismatches before
   raising (one diagnostic with every bad file, not one failure at a
   time).
3. The outcome (None on success, error string on failure) is broadcast
   via `broadcast_object_list(src=0)` so every rank raises in lock-step
   or proceeds together - no half-the-world-loaded state.
4. Single-process runs fall back to the non-distributed verify path.

### Pluggable algorithm registry

Built-in: `sha256` (default), `crc32` (faster).

Switching the active algorithm = one config field; no core code changes.
The manifest records the algorithm name, so manifests produced by one
algorithm can still be verified after the default changes.

### Config surface

```python
class CheckpointConfig(CheckpointManager.Config):
    verify_integrity: bool = False         # opt-in (see Compatibility)
    integrity_algorithm: str = "sha256"    # only used at save time
    integrity_missing_manifest: Literal["skip","warn","error"] = "warn"
    integrity_mismatch_policy: Literal["warn","error"] = "error"
    integrity_detect_extra_files: bool = True
```

- `missing_manifest`: what to do when a checkpoint written without
  `verify_integrity` is loaded (skip / warn / error).
- `mismatch_policy`: what to do when hashes don't match (warn / error).
- `detect_extra_files`: fail if the step dir contains files not in the
  manifest - catches someone planting files into a shared-NFS
  checkpoint dir. Megatron does not do this; we default it on.

## Backwards Compatibility

- **Existing checkpoints** have no manifest. With `verify_integrity=False`
  (default) the load path is untouched. If a user opts in, the default
  `integrity_missing_manifest="warn"` lets old checkpoints load with a
  WARNING - no forced re-save.
- **Manifest format** includes `algorithm`, so a future torchtitan that
  defaults to e.g. crc32 can still verify old sha256 manifests.
- **No changes to DCP file layout**. The manifest is an additional file;
  DCP loaders that don't know about it simply ignore it.

## Open Questions

1. **Default value of `verify_integrity`.** Should upstream default to
   `False` (pure opt-in, no behaviour change for existing users) or
   `True` with `integrity_missing_manifest="warn"` (safer-by-default but
   adds a one-time WARNING on first load of old checkpoints)?

2. **Manifest filename.** `_integrity_manifest.json` (prefixed with `_`
   so DCP loaders ignore it) vs `integrity.json` (Megatron-compatible
   name). Prefer the former unless there is value in cross-framework
   manifest reuse.

3. **Should `detect_extra_files` be on by default?** Catches planted
   files but may surprise users who legitimately add e.g. a `README.md`
   alongside a published checkpoint. Default on with an opt-out, or
   default off with an opt-in?

Happy to upstream the work (or a trimmed version) once the API surface
is agreed upon.

**Comment by tianyu-l** [comment]
@meetv18 @ivy-zhou

**Comment by sdmyzlp** [comment]
@chango9543 
Five new config fields for a feature that's fundamentally a boolean opt-in feels heavy. In particular:
- `integrity_mismatch_policy` with `"warn"` option: Since verification runs before `dcp.load` (per the RFC's stated order), what does "warn but proceed" mean? Neither loading anyway nor skipping the load seems like a reasonable production behavior — it seems like `"error"` is the only meaningful value.
- `integrity_missing_manifest` with `"warn"` default: This means the feature degrades to a one-time log line on every old checkpoint, which users will quickly learn to ignore.

So could this start with just `verify_integrity: bool` + `integrity_algorithm: str` (or even hardcode `sha256` like Megatron-LM), and add policy knobs later if concrete use cases arise?

**Comment by chango9543** [comment]
> [@chango9543](https://github.com/chango9543) Five new config fields for a feature that's fundamentally a boolean opt-in feels heavy. In particular:
> 
> * `integrity_mismatch_policy` with `"warn"` option: Since verification runs before `dcp.load` (per the RFC's stated order), what does "warn but proceed" mean? Neither loading anyway nor skipping the load seems like a reasonable production behavior — it seems like `"error"` is the only meaningful value.
> * `integrity_missing_manifest` with `"warn"` default: This means the feature degrades to a one-time log line on every old checkpoint, which users will quickly learn to ignore.
> 
> So could this start with just `verify_integrity: bool` + `integrity_algorithm: str` (or even hardcode `sha256` like Megatron-LM), and add policy knobs later if concrete use cases arise?

Thanks @sdmyzlp — fully agree the config surface is heavier than it needs to be. Taking your feedback, I'll trim to a single knob and hardcode the rest with sensible defaults.

### Final config surface

```python
verify_integrity: bool = False
```

That's it — one boolean, opt-in, no other policy knobs on `CheckpointManager.Config`.

### Hardcoded behaviours (and why)

The three fields I'm removing all collapse to fixed defaults:

**1. `integrity_mismatch_policy` → hardcoded `error`**

`error` is a more reasonable behavior in a production environment., so it's hardcoded. If a user genuinely wants to inspect a known-corrupt checkpoint, they can disable `verify_integrity` entirely and load it without verification.

**2. `integrity_missing_manifest` → hardcoded `skip`**

A one-time WARNING on every old checkpoint degrades to noise users learn to ignore, as you noted. `skip` keeps backwards compatibility clean.   

**3. `integrity_detect_extra_files` → hardcoded `False`**

Default-on here would surprise users who legitimately place e.g. a `README.md` alongside a published checkpoint, or stage auxiliary files next to a training checkpoint. Starting safer-the-other-way (off) avoids false positives.

### Algorithm: hardcode `sha256`, keep the registry internal

`integrity_algorithm` is also removed from the config. The default is hardcoded `sha256` — matching Megatron-LM's choice and the safest default for new checkpoints.

The pluggable algorithm registry **stays in `torchtitan/tools/checksum.py`** as an internal extension point. This keeps the design extensible: adding crc32 later is a new subclass + one `register_algorithm()` call, with zero changes to checkpoint code.

### Summary

| Field                          | Before                                    | After                                      |
| ------------------------------ | ----------------------------------------- | ------------------------------------------ |
| `verify_integrity`             | `bool = False`                            | `bool = False` (kept)                      |
| `integrity_algorithm`          | `str = "sha256"`                          | hardcoded`sha256`, registry stays internal |
| `integrity_missing_manifest`   | `Literal["skip","warn","error"] = "warn"` | hardcoded`skip`                            |
| `integrity_mismatch_policy`    | `Literal["warn","error"] = "error"`       | hardcoded`error`                           |
| `integrity_detect_extra_files` | `bool = True`                             | hardcoded`False`                           |

**Comment by ivy-zhou** [comment]
Hi there! I think this feature should be a feature of the checkpointing library itself! I mentioned in another thread that we're working on integrating torch_checkpointing into Titan, and I think this would be a great opportunity to also add this hashing feature into torch_checkpointing as well.

I made an issue to track this in torch_checkpointing (https://github.com/meta-pytorch/torch_checkpointing/issues/6) cc: @meetv18 .
